### PR TITLE
camera_info_manager: Adding public member function to manually set camera_info

### PR DESCRIPTION
**Pull request by jbohren**
See https://github.com/ros-perception/image_common/issues/19

TODO before merge:
- [x] Check ABI compatibility

### CODE REVIEW

File: camera_info_manager/src/camera_info_manager.cpp [hunk 1]
```diff
@@ -617,6 +617,24 @@ bool CameraInfoManager::setCameraName(const std::string &cname)
   return true;
 }
 
+/** Set the camera info manually
+ *
+ * @param camera_info new camera calibration data
+ *
+ * @return true if new camera info is set
+ *
+ * @post @c cam_info_ updated, if valid;
```
Comment (jbohren) [inline camera_info_manager/src/camera_info_manager.cpp:626]: I guess technically there's no validity check here. We could make sure the parameters are all reasonable, could potentially catch errors.

**Comment by jbohren** [comment]
ABI is compliant according to the abi-compliance-checker: https://rawgithub.com/jbohren/7990624/raw/eb8dd683bc63b7054194e0a85f66785e86ee3408/abi_compat_report.html

File: camera_info_manager/src/camera_info_manager.cpp (hunk 1 above)
Comment (jack-oquin) [inline camera_info_manager/src/camera_info_manager.cpp:626]: As an API, it at least gives us the option to add various checks if a good reason for them arises.

**Comment by jbohren** [comment]
Could this be merged?

**Comment by jack-oquin** [comment]
Sure. It still needs test cases.

**Comment by jack-oquin** [comment]
I just noticed this was committed to groovy-devel. You want this released to Groovy? 

I thought we were talking about Hydro (and also Indigo).

**Comment by jack-oquin** [comment]
@vrabaud: I seem to have created a minor mess here. Although backwards compatible and ABI compatible, this is still an API change. I would rather not do that in Groovy.

Please help me understand the branching scheme better. It looks like groovy-devel was last released to Groovy as 1.10.5 back in January, and that _both_ Hydro and Indigo are using hydro-devel, but Hydro is at 1.11.1 and Indigo is at 1.11.2.

Do we need or want different branches for different distros? Are we still doing new Groovy releases?

**Comment by vrabaud** [comment]
Groovy is still released so we keep fixing. The API should not be broken so that PR should be reverted.

We do a branch per distro if needed. Indigo and Hydro were the same (lately, there was a patch for Indigo only, that's why only the Indigo version got bumped).

What was the motivation for changing the API ? I can split the patch in two and only change the API in Indigo (and thus create a new branch).

**Comment by jack-oquin** [comment]
Jon's patch does not _break_ the API. It adds a new interface, which is needed for simulated camera support in Gazebo. 

I think it _could_ be released in Groovy if needed there, but extending an API in an older, stable release is not normal practice. By that standard, Hydro would also be questionable. However, when there is a good enough reason, we can make an exception.

@jbohren: do you need this released to Groovy? Hydro?

**Comment by jbohren** [comment]
> Jon's patch does not break the API. It adds a new interface, which is needed for simulated camera support in Gazebo.
> 
> I think it could be released in Groovy if needed there, but extending an API in an older, stable release is not normal practice. By that standard, Hydro would also be questionable. However, when there is a good enough reason, we can make an exception.
> 
> @jbohren: do you need this released to Groovy? Hydro?

Yeah sorry about forcing your had, @jack-oquin. I did actually mean (at this point) to merge this into Hydro. Merging into groovy now is unnecessary. I'd say we could just merge it into Indigo, but the fact that it doesn't run on 12.04 (which still blows my mind) means that I'd still have to use a fork of `image_pipeline` and `gazebo_plugins` for a while.

**Comment by jack-oquin** [comment]
Jon: I am willing to move the change to hydro-devel and release it there on both Hydro and Indigo. Most people will not be using Indigo for quite a while yet, and I would prefer to keep them both in sync until a split becomes really necessary. I don't see any harm in releasing the other hydro-devel updates to Hydro as well as Indigo.

@vrabaud: is that OK with you? Did I overlook anything?

**Comment by vrabaud** [comment]
my bad, I had not seen that setCameraInfo has been moved, not deleted. So all's good for the API.

hydro-dev is fine, and then please release for both hydro/indigo.

**Comment by jack-oquin** [comment]
Will do. Thanks for the assistance, Vincent.

**Comment by jack-oquin** [comment]
If I understand git correctly, reverting a previously merged commit could cause trouble if we were ever to merge groovy-devel into hydro-devel.

I can't think of any reason why we would ever want to do that.

**Comment by jack-oquin** [comment]
image_common 1.11.3 released to Indigo including this API change. Docs updated.

Hydro release to follow.

**Comment by jack-oquin** [comment]
Hydro release this morning.

@jbohren: for code depending on this enhancement, you should specify the `version_gte="1.11.3"` attribute.

**Pull request merged**